js/app/modules/card/constrainedThumb.js

```diff
@@ -0,0 +1,255 @@
+            },
+        ];
+        constraints.forEach((props) => this._add_constraint(props, this._layout));
+    },
+
+    vfunc_size_allocate: function (alloc) {
+        this._layout.clear_constraints();
+        this._text_layout.clear_constraints();
+        let orientation = this._get_orientation(alloc.width, alloc.height);
+
+        let show_synopsis = this._should_show_synopsis(alloc.width, alloc.height, orientation);
+
+        let show_context = !(this._should_hide_context(alloc.width, alloc.height));
+
+        let text_constraints;
+        if (orientation === Gtk.Orientation.HORIZONTAL) {
+            this._main_layout(Emeus.ConstraintAttribute.WIDTH, Emeus.ConstraintAttribute.HEIGHT);
+            text_constraints = this._get_constraints_horizontal(show_synopsis);
+            this._title_label.lines = 3;
+        } else {
+            this._main_layout(Emeus.ConstraintAttribute.HEIGHT, Emeus.ConstraintAttribute.WIDTH);
+            text_constraints = this._get_constraints_vertical(show_context);
+            this._title_label.lines = 1;
+        }
+
+        if (show_synopsis) {
+            this.set_label_or_hide(this._synopsis_label, this.model.synopsis);
+        } else {
+            this._synopsis_label.hide();
+        }
+
+        if (show_context) {
+            this._context_widget.show_all();
+        } else {
+            this._context_widget.hide();
+        }
+
+        text_constraints.forEach((props) => this._add_constraint(props, this._text_layout));
+        this.parent(alloc);
+        this.update_card_sizing_classes(alloc.height, alloc.width);
+    },
+
+    _add_constraint: function (props, layout) {
+        props.constant = props.constant || 0;
+        props.multiplier = props.multiplier || 1;
+        props.relation = props.relation || Emeus.ConstraintRelation.EQ;
+        props.source_object = props.source_object || null;
+        props.strength = props.strength || Emeus.ConstraintStrength.REQUIRED;
+        let c = new Emeus.Constraint(props);
+        layout.add_constraint(c);
+    },
+
+    _get_orientation: function (width, height) {
+        let horizontal = (width > Card.MaxSize.C && height < Card.MinSize.C) ||
+            (width > Card.MaxSize.D && height < Card.MinSize.D) ||
+            (width > Card.MaxSize.E && height < Card.MinSize.E) ||
+            (width > Card.MaxSize.F);
+        return horizontal ? Gtk.Orientation.HORIZONTAL : Gtk.Orientation.VERTICAL;
+    },
+
+    _should_show_synopsis: function (width, height, orientation) {
+        return height > Card.MaxSize.C && orientation == Gtk.Orientation.HORIZONTAL;
+    },
+
+    _should_hide_context: function (width, height) {
+        return width <= Card.MaxSize.B || (width <= Card.MaxSize.C && height <= Card.MaxSize.B);
+    },
+
+    vfunc_draw: function (cr) {
+        this.parent(cr);
+        Utils.render_border_with_arrow(this, cr);
+        cr.$dispose();  // workaround bug for not freeing cairo context
```

Comment:
Ouch, this is bad, if its true we are leaking cairo contexts we should really fix that, or find a better place for the workaround